### ISSUE TITLE
Supposed Memory Leak

### ISSUE THREAD

**Issue by FernandoDiCostanzo** [osmdroid/osmdroid#1666]
## Issue Type
Memory Leak 

## Description and/or steps/code to reproduce the problem
Testing the application with FunesDroid ( a black-box tool used to find memory leak : https://github.com/reverse-unina/FunesDroid ) noticed that some activity leaks.

FunesDroid shows a snapshot of the difference between heap memory area before experiment begin and after and so it can show remained objects in memory.
So if there some remaind objects that should means that some activity leaks and caused that remainder.

### If it's a bug, version(s) of android this affects:

All

### Version of osmdroid the issue relates to:

6.1.8

## Result of FunesDroid

[1] Here we can view wich exeperiment finds a memory leak
![risultatiFunesDroid](https://user-images.githubusercontent.com/55757804/100516954-b1e73e00-3187-11eb-9f3a-25d79ba52cf0.PNG)

[2] Here we can view the details of the exeperiment ( that found memory leaks ) an so analize the specific ativity that leaked
![osmdroid_DOC_N5_t2](https://user-images.githubusercontent.com/55757804/100516998-e6f39080-3187-11eb-826c-729b252ba92a.PNG)

By inspecting the code related to the activities it is possible to notice that they instantiate an object of the MapView type (left in heap memory [2]) which instantiates a chain of objects where at the peak there is an object of the MapTileDownoader type which, starting from a ConfigurableThreadFactory , takes care of downloading the tiles.

The solution could be to use a "Weak reference" to the ExecutorService instantiated in the MapTileModuleProviderBase class.

**Comment by monsieurtanuki** [comment]
Thank you @FernandoDiCostanzo for this analyze!

I'm not sure I understand the second table: in every line I see column "retained size" with value `0`, which sounds good, doesn't it? And I can see lines about [`MapTileArea`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/util/MapTileArea.java), which has no reference at all to `MapView`.
I wouldn't be surprised if there were leaks, but I don't manage to interpret your data. Help! ;)

Regardless, your suggestion is
> to use a "Weak reference" to the ExecutorService instantiated in the MapTileModuleProviderBase class

I'm not against the idea, but beyond the fact that I don't understand how it would solve things, there's a method named [`detach()`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/MapTileModuleProviderBase.java#L167) that is supposed to clean the instance:
```java
public void detach() {
	this.clearQueue();
	this.mExecutor.shutdown();
}
```
Assuming that method `detach()` is actually called, and maybe adding to it something like `mExecutor = null`, wouldn't we be safe?
When should it be called, then? When the activity/mapview is destroyed doesn't make sense - for instance we shouldn't stop/restart the download because of a portrait/landscape tilt.

To be investigated...

**Comment by FernandoDiCostanzo** [comment]
Hi,
about your comment:

> I'm not sure I understand the second table: in every line I see column "retained size" with value 0, which sounds good, doesn't it? And I can see lines about MapTileArea, which has no reference at all to MapView.
> I wouldn't be surprised if there were leaks, but I don't manage to interpret your data. Help! ;)

yes, rows with `retained size` equal to 0 actually indicate sequences of events that did not cause leaks, but :
FunesDroid shows principally 3 indicators:

1. `Activity leaked` : numer of leaked activity;

2. `Total shallow heap` : The shallow heap is the amount of memory consumed by an object, so the Total shallow heap it's the amount of memory consumed by all objects.

3. `Total retained size` : The retained size of the same object is the amount of heap memory that is freed when the object is garbage collected.

Since any object can be GB, the retained size is equal to 0.

Instead, about :

> And I can see lines about MapTileArea, which has no reference at all to MapView.

The activities potentially affected by leaks are:

1. `SampleWithMinimapItemizedOverlay`
2. `SampleWithTilesOverlay`
3. `SampleWithTilesOverlayAndCustomTileSource`

All of these activities instantiate an object of the `MapView `class

Trying to be as detailed as possible:
`MapView `maintains a reference and an object of the `MapTileProvideBasic `class that implements the `MapTileProviderBase interface`. In the `MapTileProvideBasic `class there is an object of type `MapTileDownloader `which extends the abstract `MapTileModuleProviderBase `class where the constructor instantiates an executor starting from a `ConfigurablePriorityThreadFactory`. In addition, `MapTileModuleProviderBase `presents an inner abstract class `TileLoader `that extends `Runnable`, which therefore allows you to instantiate threads.
Returning to the `MapTileDownloader `class we notice that it presents an abstract inner class `TileLoader `which extends the Inner abstract class of `MapTileModuleProviderBase`.
The `MapTileDownloader `class therefore has two variables of the type `TileLoader `and `TileDownloader `respectively.
The `TileLoader `class has a `downloadTile ()` method in which the `downloadTile ()` method of the `TileDownloader `class is called, which is used to download the tiles.
This occurs on dedicated threads and the leak could therefore occur when, once the `connect()` is called (following an `openConnection()`) while the operations for connecting to the remote object are carried out,
Double Orentation Change (DOC) which corresponds to a sequence of at least two changes of orientation (which cause the destruction and recreation of the activity) operations are performed and then
the activity is destroyed and rebuilt with consequent remainder of the objects that follow the chain described in the heap area as they cannot deallocated by the GC causing a permanent leak.

The solution by calling the `detach() `method seems to be ok, but : are you shure that the method is called even when a rotation occurred ?

**Comment by monsieurtanuki** [comment]
Thank you @FernandoDiCostanzo for this detailed explanation!
Now I see the general idea but still not 100% of it. I'm not an expert on leaks on Android (and I don't intend to become one anyway).

It looks like the solution would be to make sure that [`MapTileModuleProviderBase.detach()`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/MapTileModuleProviderBase.java#L167) is called by [`MapView.onDetach()`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/MapView.java#L1052), am I right?

**Comment by FernandoDiCostanzo** [comment]
Yes, the solution is ensure that the `detach()` method is also called in the `onDestroy` (which is called in the rotation cycle).

**Comment by monsieurtanuki** [comment]
It looks like `MapTileModuleProviderBase.detach()` is called by `MapView.onDetach()`:
* `MapView.onDetach()`: [`mTileProvider.detach();`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/MapView.java#L1057)
* `MapTileProviderBasic.detach()`: [`super.detach()`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/tileprovider/MapTileProviderBasic.java#L143)
* `MapTileProviderArray.detach()`: [`for (final MapTileModuleProviderBase tileProvider : mTileProviderList) {tileProvider.detach();}`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/tileprovider/MapTileProviderArray.java#L79)

Beyond that, there's no `onDestroy()` method for `View` it seems, just `onDetachedFromWindow()`, where we call `onDetach` most of the time (cf. #1247, #1274):
```java
@Override
protected void onDetachedFromWindow() {
	if (mDestroyModeOnDetach) {
		onDetach();
	}
	super.onDetachedFromWindow();
}
```

**Comment by mikebravoyes** [comment]
@monsieurtanuki @FernandoDiCostanzo 

I have a similar problem with a memory leak in my app. That has several activities, one of them a map activity, where I show a background `MAPNIK` with some `GroundOverlay`s with weather maps. 

![mapView+GroundOverlays](https://user-images.githubusercontent.com/6436590/107342750-e7ca6c00-6ac0-11eb-81cd-3a6bfc61f994.png)

When I switch between the activities - or when I rotate the device from portrait to landscape and back - the memory consumption is increasing each time, when the map is rebuilt. Sooner or later (depending upon the device memory) I will get an `OutOfMemoryError`.

I tried to introduce 

    @Override
    protected void onStop ()
    {
        mapView.onDetach(); 
        super.onStop();
    }  // onStop

But that didn't bring any improvement.

`mapView.onDetachedFromWindow()` was not called during my tests.

So what is the outcome of your discussion? Is there a solution or workaround for that memory leak problem?

**Comment by monsieurtanuki** [comment]
@mikebravoyes The weather maps look much more complex (and heavy) than the standard MAPNIK tiles: that would explain why you manage to get an OOM exception quickly.
Could you create a simple sample app with just the MAPNIK as background and all your GroundOverlays on top, and reproduce the bug?

**Comment by mikebravoyes** [comment]
@monsieurtanuki
OK, I have trimmed the app to the absolute minimum - still a whole bunch of files. It's called "MapDemo" and I have attached the apk file (zipped) and the complete Android Studio project including source code. Please concentrate on the files `Act06_Chart.java` (lines #1050ff) and `WeatherOverlay.java` (lines #200ff) and ignore the rest.

[app-debug.zip](https://github.com/osmdroid/osmdroid/files/5958998/app-debug.zip)
[MapDemo.zip](https://github.com/osmdroid/osmdroid/files/5958999/MapDemo.zip)

When you start the app for the first time, you have to grant the permissions for "Files and media". Otherwise the map will not work and the app cannot store the last settings.

**Comment by monsieurtanuki** [comment]
@mikebravoyes The idea is to reproduce the bug with the simplest app (about 50 lines of code), not with 30 classes up to 2000 lines of code each.
Btw I don't code on osmdroid anymore. I read code and suggest solution and bug fixes.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
I thought you wanted an executable app - a misunderstanding.
Now here is the extract:

In `Act06_Chart.java`
```
WeatherOverlay weatherOverlay = new WeatherOverlay(this, mapView, mapOverlays, wdStack);
mapView.getOverlays().add(0, weatherOverlay);
```

and in `WeatherOverlay.java`
```
UrlString = "https://en.sat24.com/image?type=km&region=de";
UrlName = "Germany";

// load image from website
URL url = new URL(UrlString);
Connection  connection = (HttpsURLConnection) url.openConnection();
InputStream inputStream = connection.getInputStream();

// read the image
Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
inputStream.close();
```

**Comment by monsieurtanuki** [comment]
@mikebravoyes I'm afraid I cannot be helpful there.
My suggestion: create the simplest app and try to make it crash the way you expect it to crash. If you cannot make it crash, maybe the problem is in the rest of your code. If it does crash though, then there's less fat around the code and you can more easily focus on the apparently important steps: the Activity destroy / view detach operations.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
For your information: The solution to avoid memory leaks was the following code in the activity `class Act06_Chart`

```
@Override
protected void onStop ()
{
    mapView.getOverlays().clear();
    mapView.onDetach(); 

    super.onStop();
}  // onStop
```

Thanks for your help.

**Comment by monsieurtanuki** [comment]
Thank you @mikebravoyes for your feedback.
I'm a bit surprised though: if I read the code it looks like `mapView.onDetach()` eventually calls something like `mapView.getOverlays().clear()`.
Wouldn't calling just `mapView.onDetach()` be enough?

**Comment by mikebravoyes** [comment]
According to Android Studio, the `mapView.onDetach()` contains this code:

```
/**
 * destroys the map view, all references to listeners, all overlays, etc
 */
public void onDetach() {
	this.getOverlayManager().onDetach(this);
	mTileProvider.detach();
	if (mZoomController != null) {
		mZoomController.onDetach();
	}

	//https://github.com/osmdroid/osmdroid/issues/390
	if (mTileRequestCompleteHandler instanceof SimpleInvalidationHandler) {
		((SimpleInvalidationHandler) mTileRequestCompleteHandler).destroy();
	}
	mTileRequestCompleteHandler=null;
	if (mProjection!=null)
		mProjection.detach();
	mProjection=null;
	mRepository.onDetach();
	mListners.clear();
}
```

And the first line `this.getOverlayManager().onDetach(this);` invokes
`
void onDetach(MapView pMapView);
`
which is empty, it doesn't do anything.

**Comment by monsieurtanuki** [comment]
@mikebravoyes This is java here: `void onDetach(MapView pMapView);` does not mean that we do nothing, it means it's not implemented (like in an `interface` or an `abstract class`).

Actually in `MapView` the instance of `OverlayManager` is a `DefaultOverlayManager`, which implements `onDetach` [this way](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/DefaultOverlayManager.java#L188-L198):
```java
@Override
public void onDetach(final MapView pMapView) {
    if (mTilesOverlay != null) {
        mTilesOverlay.onDetach(pMapView);
    }

    for (final Overlay overlay : this.overlaysReversed()) {
        overlay.onDetach(pMapView);
    }
    this.clear();
}
```
As it ends with `clear()`, I assume that your call to `mapView.getOverlays().clear()` is useless - maybe even problematic, as you don't call `onDetach()` on each `Overlay`.

@mikebravoyes Could you please double-check that the overlays are getting cleared with only `mapView.onDetach()` - something like that:
```java
@Override
protected void onStop ()
{
    mapView.onDetach(); 
    Log.e("test", "size=" + mapView.getOverlays().size());
    super.onStop();
}
```

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
You are right, the version with the single `mapView.onDetach();` works as well, with size=0.
I have tested that two weeks ago and had found that it did not work (as stated above). To be honest, I don't understand that. Maybe I have fixed another memory leak accidentally by various modifications (embarrassing...).

Do you really think, that the redundant `mapView.getOverlays().clear();` may cause any trouble, even if it's not required?

**Comment by monsieurtanuki** [comment]
@mikebravoyes I'm sure that calling `mapView.getOverlays().clear()` _before_ `mapView.onDetach()` can be dangerous (memory-wise), because it does less things and prevents `onDetach` to be called on each `Overlay`.
If you really want to add this line, add it _after_ `mapView.onDetach()`, it will be pointless but harmless here.

Regarding the bug fix, as you were able to reproduce the bug every time you wanted, if you cannot reproduce it anymore let's say it's good news! We may never know what exactly you inadvertently fixed, but your code works now, and so does osmdroid when `mapView.onDetach()` is called appropriately.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
OK, thanks for your helpful support.

**Comment by FernandoDiCostanzo** [comment]
Hi @monsieurtanuki,
Have you fixed the memory leak or it still exists ?

**Comment by mikebravoyes** [comment]
@FernandoDiCostanzo
Let me answer for @monsieurtanuki 
The modification which fixed the memory leak was the following code in the activity:

```
@Override
protected void onStop ()
{
    mapView.onDetach(); 

    super.onStop();
}
```
All the overlays were cleared by that. It's as simple as that.

**Comment by FernandoDiCostanzo** [comment]
Perfect.
Can i have the apk on wich that modifics was applied ? 
I want to test it with FunesDroid another time to be shure of the fix.

**Comment by mikebravoyes** [comment]
@FernandoDiCostanzo 
1.) I observed the memory utilisation with the profiler of Android Studio and found that the leak is gone.
2.) What do you mean with apk? The app is available in the Google Play Store:
- [Lite Version](https://play.google.com/store/apps/details?id=net.braun_home.sensorrecording.lite) (free of charge, data recording limited to 1 minute)
- [Pro Version](https://play.google.com/store/apps/details?id=net.braun_home.sensorrecording.pro) (1,89 €, full functionality)

If you are talking about the source code: That is not published.

**Comment by monsieurtanuki** [comment]
@FernandoDiCostanzo If you use osmdroid, the fix seems to be an override of your activity or fragment (as @mikebravoyes mentioned earlier):
```java
@Override
protected void onStop () {
    mapView.onDetach(); 
    super.onStop();
}
```

Everything (`mapView.onDetach()`) is already available; you just have to override `onStop` _in your own code_ (we cannot do that for you).
